trading_calendars/exchange_calendar_xpsx.py

```diff
@@ -0,0 +1,394 @@
+#
+# Copyright 2019 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import time
+from itertools import chain
+
+import pandas as pd
+from pandas.tseries.holiday import Holiday
+from pytz import timezone
+
+from .common_holidays import european_labour_day
+from .trading_calendar import HolidayCalendar, TradingCalendar
+
+
+class XPSXExchangeCalendar(TradingCalendar):
+    """
+    Exchange calendar for the Pakistan Stock Exchange (XPSX).
+
+    Open Time:
+    Close Time:
+
+    Regularly-Observed Holidays:
+    - Kashmir Day (Feb 5)
+    - Pakistan Day (Mar 23)
+    - Labour Day (May 1)
+    - Juma-Tul-Wida (last Friday of Ramadan)
+    - Eil-ul-Fitr (1st-3rd Shawwal)
+    - Eid-ul-Azha (10th-11th Zil-Hajj)
+    - Independence Day (Aug 14)
+    - Ashura (9th & 10th Muharram)
+    - Eid Milad-un-Nabi (12th Rabi-ul-Awal)
+    - Birthday of Quaid-e-Azam & Christmas (Dec 25)
+
+    Formerly-Observed Holidays:
+    - Iqbal Day (Nov 9; ended in 2012)
```

Comment:
If you think it's worth it, may be good to add a quick test checking this date before/after 2012.